File: packages/kirie/native/ios/Kirie/Sources/Kirie/KirieURLResolver.swift
import Foundation

struct KirieResolvedURL {
    let url: URL
    let readAccessURL: URL?
}

enum KirieURLResolver {
    private static let resSchemePrefix = "res://"

    static func resolveForWebView(_ urlString: String) throws -> KirieResolvedURL {
        guard urlString.lowercased().hasPrefix(resSchemePrefix) else {
            guard let url = URL(string: urlString) else {
                throw KirieURLResolverError.invalidURL(urlString)
            }

            return KirieResolvedURL(url: url, readAccessURL: nil)
        }

        return try resolveResourceURL(urlString)
    }

    private static func resolveResourceURL(_ urlString: String) throws -> KirieResolvedURL {
        let suffix = String(urlString.dropFirst(resSchemePrefix.count))
        let pathEnd = suffix.firstIndex { character in
            character == "?" || character == "#"
        } ?? suffix.endIndex
        let encodedPath = String(suffix[..<pathEnd])
        let queryAndFragment = String(suffix[pathEnd...])

        guard !encodedPath.isEmpty else {
            throw KirieURLResolverError.emptyResourcePath
        }

        guard hasValidPercentEncoding(encodedPath) else {
            throw KirieURLResolverError.invalidPercentEncoding(urlString)
        }

        let decodedPath = encodedPath.removingPercentEncoding?.trimmingCharacters(in: CharacterSet(charactersIn: "/")) ?? ""
        guard !decodedPath.isEmpty else {
            throw KirieURLResolverError.emptyResourcePath
        }

        guard !decodedPath.contains("\0"), !containsPathTraversal(decodedPath) else {
            throw KirieURLResolverError.unsafeResourcePath(urlString)
        }

        guard let bundleResourceURL = Bundle.main.resourceURL else {
            throw KirieURLResolverError.missingBundleResourceURL
        }

        let assetPath = resolveAssetPath(decodedPath)
        let candidates = resourceCandidates(for: assetPath, bundleResourceURL: bundleResourceURL)
        guard let resolvedResource = candidates.first(where: { FileManager.default.fileExists(atPath: $0.fileURL.path) }) else {
            throw KirieURLResolverError.resourceNotFound(
                assetPath,
                candidates.map { "\($0.assetPath) at \($0.fileURL.path)" }
            )
        }

        let navigableURL = try appendQueryAndFragment(queryAndFragment, to: resolvedResource.fileURL)
        let readAccessURL = readAccessRoot(for: resolvedResource.assetPath, bundleResourceURL: bundleResourceURL)
        return KirieResolvedURL(url: navigableURL, readAccessURL: readAccessURL)
    }

    private static func appendQueryAndFragment(_ queryAndFragment: String, to url: URL) throws -> URL {
        guard !queryAndFragment.isEmpty else {
            return url
        }

        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        if queryAndFragment.hasPrefix("?") {
            let tail = queryAndFragment.dropFirst()
            if let fragmentStart = tail.firstIndex(of: "#") {
                components?.percentEncodedQuery = String(tail[..<fragmentStart])
                components?.percentEncodedFragment = String(tail[tail.index(after: fragmentStart)...])
            } else {
                components?.percentEncodedQuery = String(tail)
            }
        } else if queryAndFragment.hasPrefix("#") {
            components?.percentEncodedFragment = String(queryAndFragment.dropFirst())
        }

        guard let resolvedURL = components?.url else {
            throw KirieURLResolverError.invalidURL(url.absoluteString + queryAndFragment)
        }

        return resolvedURL
    }

    private static func readAccessRoot(for assetPath: String, bundleResourceURL: URL) -> URL {
        let segments = assetPath.split(separator: "/", omittingEmptySubsequences: true)

        guard segments.count > 1, let firstSegment = segments.first, !firstSegment.contains(".") else {
            return bundleResourceURL
        }

        return bundleResourceURL.appendingPathComponent(String(firstSegment), isDirectory: true)
    }

    private static func resourceCandidates(for assetPath: String, bundleResourceURL: URL) -> [KirieBundleResourceCandidate] {
        var candidates = [
            KirieBundleResourceCandidate(assetPath: assetPath, fileURL: bundleResourceURL.appendingPathComponent(assetPath, isDirectory: false)),
        ]

        if assetPath.hasPrefix("web/") {
            candidates.append(
                KirieBundleResourceCandidate(
                    assetPath: String(assetPath.dropFirst("web/".count)),
                    fileURL: bundleResourceURL.appendingPathComponent(String(assetPath.dropFirst("web/".count)), isDirectory: false)
                )
            )
        }

        if let executableName = Bundle.main.object(forInfoDictionaryKey: "CFBundleExecutable") as? String {
            candidates.append(
                KirieBundleResourceCandidate(
                    assetPath: "\(executableName)/\(assetPath)",
                    fileURL: bundleResourceURL
                        .appendingPathComponent(executableName, isDirectory: true)
                        .appendingPathComponent(assetPath, isDirectory: false)
                )
            )
        }

        return candidates
    }

    private static func resolveAssetPath(_ path: String) -> String {
        let trimmedPath = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let lastSegment = trimmedPath.split(separator: "/").last.map(String.init) ?? ""

        if path.hasSuffix("/") || !lastSegment.contains(".") {
            return "\(trimmedPath)/index.html"
        }

        return trimmedPath
    }

    private static func hasValidPercentEncoding(_ value: String) -> Bool {
        var index = value.startIndex

        while index < value.endIndex {
            guard value[index] == "%" else {
                index = value.index(after: index)
                continue
            }

            let firstHexIndex = value.index(after: index)
            guard firstHexIndex < value.endIndex else {
                return false
            }

            let secondHexIndex = value.index(after: firstHexIndex)
            guard secondHexIndex < value.endIndex else {
                return false
            }

            guard value[firstHexIndex].isHexDigit, value[secondHexIndex].isHexDigit else {
                return false
            }

            index = value.index(after: secondHexIndex)
        }

        return true
    }

    private static func containsPathTraversal(_ path: String) -> Bool {
        path.replacingOccurrences(of: "\\", with: "/")
            .split(separator: "/", omittingEmptySubsequences: false)
            .contains("..")
    }
}

private struct KirieBundleResourceCandidate {
    let assetPath: String
    let fileURL: URL
}

enum KirieURLResolverError: LocalizedError {
    case emptyResourcePath
    case invalidPercentEncoding(String)
    case invalidURL(String)
    case missingBundleResourceURL
    case resourceNotFound(String, [String])
    case unsafeResourcePath(String)

    var errorDescription: String? {
        switch self {
        case .emptyResourcePath:
            return "Cannot load empty res:// asset path"
        case let .invalidPercentEncoding(url):
            return "Cannot load res:// URL with invalid percent encoding: \(url)"
        case let .invalidURL(url):
            return "Cannot load invalid URL: \(url)"
        case .missingBundleResourceURL:
            return "Cannot load res:// URL because the app bundle resource URL is unavailable"
        case let .resourceNotFound(assetPath, candidates):
            return "Cannot load res:// asset because it was not found in the app bundle: \(assetPath); checked \(candidates.joined(separator: ", "))"
        case let .unsafeResourcePath(url):
            return "Cannot load unsafe res:// asset path: \(url)"
        }
    }
}

File: packages/kirie/native/android/plugin/src/main/java/ai/moeru/kirie/android/KirieUrlResolver.kt
package ai.moeru.kirie.android

import android.net.Uri

object KirieUrlResolver {
    private const val RES_SCHEME_PREFIX = "res://"
    private const val KIRIE_ASSET_SCHEME = "https"
    private const val KIRIE_ASSET_HOST = "res.kirie.invalid"
    private const val KIRIE_ASSET_BASE_URL = "$KIRIE_ASSET_SCHEME://$KIRIE_ASSET_HOST/"

    fun resolveForWebView(url: String): String {
        if (!url.startsWith(RES_SCHEME_PREFIX, ignoreCase = true)) {
            return url
        }

        val suffix = url.substring(RES_SCHEME_PREFIX.length)
        val pathEnd = suffix.indexOfAny(charArrayOf('?', '#')).takeIf { it != -1 } ?: suffix.length
        val encodedPath = suffix.take(pathEnd)
        val queryAndFragment = suffix.drop(pathEnd)

        require(encodedPath.isNotBlank()) { "Cannot load empty res:// asset path" }
        require(hasValidPercentEncoding(encodedPath)) { "Cannot load res:// URL with invalid percent encoding: $url" }

        val decodedPath = Uri.decode(encodedPath).trimStart('/')
        require(decodedPath.isNotBlank()) { "Cannot load empty res:// asset path" }
        require(decodedPath.indexOf('\u0000') == -1 && !containsPathTraversal(decodedPath)) {
            "Cannot load unsafe res:// asset path: $url"
        }

        val assetPath = resolveAssetPath(decodedPath)
        return KIRIE_ASSET_BASE_URL + Uri.encode(assetPath, "/") + queryAndFragment
    }

    fun isResolvedAssetUrl(uri: Uri): Boolean =
        KIRIE_ASSET_SCHEME.equals(uri.scheme, ignoreCase = true) &&
            KIRIE_ASSET_HOST.equals(uri.host, ignoreCase = true)

    fun assetPathFromResolvedUrl(uri: Uri): String {
        require(isResolvedAssetUrl(uri)) { "Not a Kirie asset URL: $uri" }

        val encodedPath = uri.encodedPath?.trimStart('/') ?: ""
        require(encodedPath.isNotBlank()) { "Cannot load empty Kirie asset path" }
        require(hasValidPercentEncoding(encodedPath)) { "Cannot load Kirie asset URL with invalid percent encoding: $uri" }

        val decodedPath = Uri.decode(encodedPath)
        require(decodedPath.indexOf('\u0000') == -1 && !containsPathTraversal(decodedPath)) {
            "Cannot load unsafe Kirie asset path: $uri"
        }

        return decodedPath
    }

    private fun hasValidPercentEncoding(value: String): Boolean {
        var index = 0
        while (index < value.length) {
            if (value[index] != '%') {
                index += 1
                continue
            }

            if (
                index + 2 >= value.length ||
                !value[index + 1].isHexDigit() ||
                !value[index + 2].isHexDigit()
            ) {
                return false
            }

            index += 3
        }

        return true
    }

    private fun containsPathTraversal(path: String): Boolean =
        path
            .replace('\\', '/')
            .split('/')
            .any { segment -> segment == ".." }

    private fun resolveAssetPath(path: String): String {
        val trimmedPath = path.trimEnd('/')
        val lastSegment = trimmedPath.substringAfterLast('/')

        return if (path.endsWith('/') || !lastSegment.contains('.')) {
            "$trimmedPath/index.html"
        } else {
            path
        }
    }

    private fun Char.isHexDigit(): Boolean = this in '0'..'9' || this in 'a'..'f' || this in 'A'..'F'
}
